### ISSUE TITLE
Improve tree-shaking in bundles for new installation methods

### ISSUE THREAD

**Issue by filipsobol** [ckeditor/ckeditor5#16292]
When working on the [new installation methods (NIM)](https://github.com/ckeditor/ckeditor5/issues/15502), we noticed that the bundle size in the consuming projects was larger than we expected. Investigation showed that there was a lot of code in various core and commercial packages that was not properly tree-shaken.

Although we noticed this while working on NIM, this is not a new problem. Because in the old installation methods the plugins are imported directly from their respective packages, it was much less likely that the code that wasn't properly tree-shaken would end up unused. However, in NIM, all core and commercial packages are imported from the `ckeditor5` and `ckeditor5-premium-feature` packages. This means that pulling code from one plugin will also pull side-effects from all other plugins in these packages.

How significant is this issue? When using the `0.0.0-nightly-20240426.0` build, the following imports will result in a bundle size of 1390.65 kB in Rollup and esbuild (minified, not gzipped):

```js
import 'ckeditor5';
import 'ckeditor5-premium-features';
```

In a perfect world, the bundle size should be 0.0 kB because we didn't import anything specific. However, about 550 kB of that 1390.65 kB are essential packages (`engine`, `ui`, `core`, `utils`, `watchdog`) that are used by every plugin. While we could try to make this part tree-shakable, that's not the main focus of this task. The main focus now is to make the other 840 kB tree-shakable.

Our initial tests show that we can reduce the bundle size to 738 kB (or 182 kB minzipped) without major code changes, with the potential to reduce the size by another 150 kB. This would be a 58% reduction from the current state.

## Problems

Below are the main problems that we identified.

### Static methods and fields

In our codebase, we make heavy use of static methods and fields. While support for static public methods was introduced in ES2015, support for static private methods and static fields (public and private) was added in ES2022.

Because support for all four combinations has long been available in TypeScript, it and other bundlers had to transpile them to support the older browser. Unfortunately, these workarounds make the output code less tree-shakable. For a simple example, consider the following code:

```ts
export class Test {
  public static field = 123;

  public static method() {
    return 123;
  }
}
```

Since public static methods are well supported, the `public static method()` is transpiled to `static method()`. However, the `public static field` is transpiled to code that cannot be tree-shaken, and the output varies between transpilers. See the following links for TypeScript, esbuild, and SWC outputs:

* [esbuild](https://hyrious.me/esbuild-repl/?version=0.20.2&b=e%00entry.ts%00export+class+Test+%7B%0A++public+static+field+%3D+123%3B%0A%0A++public+static+method%28%29+%7B%0A++++return+123%3B%0A++%7D%0A%7D&o=--format%3Desm+--target%3Des2019).
* [SWC](https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAA0utKMgvKlFIzkksLlYISS0uUajmUlAoKE3KyUxWKC5JLAFSaZmpOSkKtgqGRsbWXBiyuaklGfkpGppgjQoKRaklpUV5ELUKCrVctQBx1NzrYgAAAA%3D%3D&config=H4sIAAAAAAAAA2WQyw6CQAxF934F6dqFujDRtVs%2BohkKwcyDtCWREP7d4TEjxF3b257edjwVBbzFwLMYYxiTDlmIcx4rMnjFT6yADh2J4bZTOCdVZZZqtEK5VpEJjBpYoqTc0yJMqw6K3JDOPJLb5frYWGBDEDqywLW%2BrYe9GxNcxyTyt9Shb2yeP25k9FIHdntQNlmSYoWKm9cMtNSgGV6p7XDKaYODC1W%2FbN3eN79oPe0Ov6Z0RXYMrZRpcqFOX7j%2FIqaKAQAA).
* [TypeScript](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBAYwDYEMDOa4BVhvgbwCg44wBXAIyQEsE48UZa4Aza4JAEzgF44BGAEwBmANyFipSjToMmdALbAYACwicAFAEo4REiSjKyUAHYCR4kgF9CVoA).

Change the output target to ES2022 to see how the workarounds disappears to something that can be easily analyzed and tree-shaken.

We can't change the target in the entire codebase because that would break support for webpack 4 (that's why we only target ES2019 for now), but we can [change the target only in our new CLI tool](https://github.com/ckeditor/ckeditor5-dev/blob/2b8cb5f3ce98f0501b14054f76b0a052cf16126d/packages/ckeditor5-dev-build-tools/src/config.ts#L178) to only affect new installation methods.

Before we can do that, however, we need to fix the problem described in #14703 and #16386, because part of our code is not valid and only works because of the transpilation.

### Mixins

We use mixins in our codebase to share code between classes. However, the current implementation of mixins in TypeScript is not tree-shakable. Consider the following code:

```ts
class FileLoader extends ObservableMixin() {}
```

When statically analyzing the code, it's impossible for bundlers to know what `ObservableMixin()` will return and if it contains side effects. This makes (some) classes that extend mixins not tree-shakable.

This can be easily fixed by adding the [`#__PURE__`](https://rollupjs.org/configuration-options/#pure) comment to all mixin calls, like this:

```ts
class FileLoader extends /* #__PURE__ */ ObservableMixin() {}
```

Alternatively (or even preferably), we could try annotating the mixin functions with [`#__NO_SIDE_EFFECTS__`](https://rollupjs.org/configuration-options/#no-side-effects), but we need to make sure that this comment works when using webpack, Vite, and esbuild.

### Other problems

- Remove code added for [backward compatibility with `mix`](https://github.com/search?q=repo%3Ackeditor%2Fckeditor5+%22Backward+compatibility+with+%60mix%60%22&type=code).
- Make it so that protobufjs and compiled messages can be tree-shaken when real-time plugins are not used.
- (Lower priority) To get the types for the `is` methods right, we used some [workarounds](https://github.com/search?q=repo%3Ackeditor%2Fckeditor5+%22.prototype.is+%3D+function%22&type=code) which are not tree-shakable. We should move the `is` implementations to static methods without breaking the API / types.
- (Lower priority) The way we use [`turndown`](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-markdown-gfm/src/html2markdown/html2markdown.ts) and [`marked`](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-markdown-gfm/src/markdown2html/markdown2html.ts) makes the bundlers unable to tree-shake them. We should refactor the code to fix it.
- (Lower priority) Replace enums with regular JavaScript objects where possible.

### (One time) Setup repository

1. Clone this repository.
2. Create the `external` folder in the root of the repository.
3. Clone the [ckeditor5-dev](https://github.com/ckeditor/ckeditor5-dev) repository into the `external` folder.
4. Open the `external/ckeditor5-dev/packages/ckeditor5-dev-build-tools/src/config.ts` file and add the `experimentalLogSideEffects: true` property to the object returned from the `getRollupConfig` function.
5. Open the `external/ckeditor5-dev/packages/ckeditor5-dev-build-tools/src/build.ts` file and add `preserveModules: true` to the object passed to the `build.write` in the `build` function (not the `generateUmdBuild` function).
6. Run `yarn reinstall` in the root of the `ckeditor5` repository. This should install all dependencies and build the changes we made in the `ckeditor5-dev` repository.

### (One time) Setup Verdaccio

1. [Install Verdaccio](https://verdaccio.org/docs/installation/).
2. Create `.npmrc` file and add `registry=http://localhost:4873/` to it.
3. Run `npm adduser --registry http://localhost:4873/`. Make sure to name the user `ckeditor` (this is required by one of our internal scripts).

### (One time) Setup test repository

1. Clone the [new-installation-methods](https://github.com/ckeditor/new-installation-methods) repository.
2. In the `vite` directory create the `.npmrc` file and add `registry=http://localhost:4873/` to it.
3. Run `npm install` in the `vite` directory.
4. Change the contents of `vite/src/main.ts` to:
    ```ts
    import 'ckeditor5';
    import 'ckeditor5-premium-features';
    ```

### Test changes

In the `ckeditor5` repository:
1. Make the changes in the code.
2. Stage the changes (`git add .`).
3. Run `npm run release:prepare-packages -- --nightly` in the root of the `ckeditor5` repository (you can comment out some of the steps in the `scripts/release/preparepackages.js` file to speed up the process).
6. Run `npm run release:publish-packages -- --nightly` to publish the changes to Verdaccio.
7. Revert all unstaged changes (`git restore .`).

In the `new-installation-methods` repository:
1. Reinstall the dependencies in the `vite` directory.
2. Run `npm run build` to see the changes in the bundle size.
3. Run the following command and check the bundle size:
    ```sh
    npx esbuild ./src/main.ts \
      --bundle \
      --minify \
      --format=esm \
      --outfile=custom/bundle.js \
      --metafile=meta.json \
      --legal-comments=none \
      --target=es2022
    ```
4.  Upload the `meta.json` file to the [esbuild analyzer](https://esbuild.github.io/analyze/) for a more detailed analysis.

**Comment by filipsobol** [comment]
| Name                                      | Rollup \| side-effect | Rollup \| build | esbuild \| Side-effects | esbuild \| Build | Version                      | Associated PRs |
|-------------------------------------------|-----------------------|-----------------|-------------------------|------------------|------------------------------|----------------|
| Baseline                                  | 1377                  | 1819            | 1412                    | 1779             | 0.0.0-nightly-20240426.0 |                |
| `/* #__PURE__ */` comments in mixin calls | 1374                  | 1820            | 1396                    | 1773             | 0.0.0-nightly-20240509.0 |                |
| Bumping target to ES2022                  | 679                   | 1537            | 1295                    | 1715             | 0.0.0-nightly-20240510.0 |                |

**Comment by filipsobol** [comment]
This task is complete. I created a new one describing how we can further improve the bundle size.

https://github.com/ckeditor/ckeditor5/issues/16395
